### PR TITLE
plutus-contract: Better test output

### PR DESCRIPTION
**Pull request by j-mueller**
* Add a log to the `TracePredicate` type
* Sample output:

grafik

I had a look around on hackage, the `prednote` package seems quite promising but it needs some maintenance (doesn't build with current GHC) so I decided to just use a newtype.

### CODE REVIEW

File: plutus-contract/src/Language/Plutus/Contract/Test.hs [hunk 1]
```diff
@@ -52,7 +56,19 @@ import qualified Wallet.Emulator                       as EM
 
 import           Language.Plutus.Contract.Trace        as X
 
-type TracePredicate a = InitialDistribution -> Predicate (ContractTraceResult a)
+newtype PredF f a = PredF { unPredF :: a -> f Bool }
+
+instance Contravariant (PredF f) where
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Test.hs]: Is this not derivable? Presumably there's an instance for `->`

File: plutus-contract/src/Language/Plutus/Contract/Test.hs [hunk 2]
```diff
@@ -52,7 +56,19 @@ import qualified Wallet.Emulator                       as EM
...
+    contramap f (PredF p) = PredF (p . f)
+
+instance Applicative f => Semigroup (PredF f a) where
+    l <> r = PredF $ \a -> (&&) <$> unPredF l a <*> unPredF r a
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Test.hs]: I'm not sure about this since it's privileging the AND monoid for booleans...

Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: Do you actually use this?

File: plutus-contract/src/Language/Plutus/Contract/Test.hs [hunk 3]
```diff
@@ -76,65 +92,144 @@ checkPredicate
...
+walletState w flt nm = PredF $ \(_, r) -> do
+    let ws = view (at w) $ EM._walletStates $  _ctrEmulatorState r
+    case ws of
+        Nothing -> writer (False, Seq.singleton ("Wallet state of '" <> show w <> "' not found"))
+        Just st ->
+            if flt st
+            then pure True
+            else do
+                tellSeq ["Wallet state of " <> show w <> ":", show st, "Fails '" <> nm <> "'"]
+                pure False
```
Comment (michaelpj) [inline plutus-contract/src/Language/Plutus/Contract/Test.hs]: You're a little inconsistent about when you use short `do` blocks vs `writer`. I'd probably use the `do` blocks everywhere tbh.

File: plutus-contract/src/Language/Plutus/Contract/Test.hs (hunk 2 above)
Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: yes I actually use it a lot in the tests, ANDing the predicates together there is the default mode in my opinion

Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: Actually, you know what the better solution is? Lattice!

File: plutus-contract/src/Language/Plutus/Contract/Test.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: There isn't an instance for `->` :(

Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: Oh of course, it has to be the last argument :roll_eyes:

There is `Contravariant.Op`, so you could do deriving via that.

File: plutus-contract/src/Language/Plutus/Contract/Test.hs (hunk 2 above)
Comment (michaelpj) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: But fine for now. I'm just going to bite the bullet and write a lattice class.

Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: > Actually, you know what the better solution is? Lattice!

As usual ;)

File: plutus-contract/src/Language/Plutus/Contract/Test.hs (hunk 1 above)
Comment (j-mueller) [reply plutus-contract/src/Language/Plutus/Contract/Test.hs]: Ah yeah I was looking for somthing like that, thanks.

**Comment by michaelpj** [comment]
This is the known failure, merging.

**Pull request merged**